lib/mayaUsd/commands/layerEditorCommand.cpp

```diff
@@ -547,32 +548,116 @@ class BackupLayerBase : public BaseCmd
+    void backupLayer(SdfLayerHandle layer)
+    {
+        if (!layer)
+            return;
+
+        if (layer->IsDirty()) {
+            _backupLayer = SdfLayer::CreateAnonymous();
+            _backupLayer->TransferContent(layer);
+        }
+    }
+
+    void restoreLayer(SdfLayerHandle layer)
+    {
+        if (!layer)
+            return;
+
+        if (_backupLayer) {
             layer->TransferContent(_backupLayer);
             _backupLayer = nullptr;
-            releaseSubLayers();
+        } else {
+            layer->Reload();
         }
-        return true;
     }
 
+    // Backup and restore edit targets of stages that were targeting the sub-layers
+    // of the cleared layer to support undo and redo.
+    void backupEditTargets(SdfLayerHandle layer)
+    {
+        _editTargetBackups.clear();
+
+        if (!layer)
+            return;
+
+        const UsdMayaStageCache::Caches& caches = UsdMayaStageCache::GetAllCaches();
+        for (const PXR_NS::UsdStageCache& cache : caches) {
+            const std::vector<UsdStageRefPtr> stages = cache.GetAllStages();
+            for (const PXR_NS::UsdStageRefPtr& stage : stages) {
+                if (!stage)
+                    continue;
+                const PXR_NS::UsdEditTarget target = stage->GetEditTarget();
+                // Note: this is the check that UsdStage::SetTargetLayer would do
+                //       which is how we would detect that the edit target is now
+                //       invalid. Unfortunately, there is no USD function to check
+                //       if an edit target is valid outside of trying to set it as
+                //       the edit target, but we would not want to set it. (Also,
+                //       knowing if the stage checks that the edit target is already
+                //       set to the same target before validating it is an implementation
+                //       detail that we would raher not rely on.)
+                if (stage->HasLocalLayer(target.GetLayer()))
+                    continue;
+                _editTargetBackups[stage] = target;
+
+                // Set a valid target. The only layer we can count on is the root
+                // layer, so set the target to that.
+                stage->SetEditTarget(stage->GetRootLayer());
+            }
+        }
+    }
+
+    void restoreEditTargets()
+    {
+        for (const auto& weakStageAndTarget : _editTargetBackups) {
+            const PXR_NS::UsdStageRefPtr stage = weakStageAndTarget.first;
+            if (!stage)
+                continue;
+
+            PXR_NS::UsdEditTarget target = weakStageAndTarget.second;
+            stage->SetEditTarget(target);
+        }
+    }
+
+    // Edit targets that were made invalid after the layer was cleared.
+    // The stages are kept with weak pointer to avoid forcing to stay valid.
+    using EditTargetBackups = std::map<PXR_NS::UsdStagePtr, PXR_NS::UsdEditTarget>;
```

Comment:
Putting here as a reference: I talked to Pierre and the map is for the scenario that multiple stages share the same layer and we need to make sure when we restore edit target, we restore to the correct stage